pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1433,41 +1467,69 @@ func (og *operationGenerator) GenerateExpandVolumeFSWithoutUnmountingFunc(
 		return volumetypes.GeneratedOperations{}, volumeToMount.GenerateErrorDetailed("VolumeFSResize.FindPluginBySpec failed", err)
 	}
 
-	attachableVolumePlugin, err :=
-		og.volumePluginMgr.FindAttachablePluginBySpec(volumeToMount.VolumeSpec)
-	if err != nil || attachableVolumePlugin == nil {
-		if attachableVolumePlugin == nil {
-			err = fmt.Errorf("AttachableVolumePlugin is nil")
-		}
-		return volumetypes.GeneratedOperations{}, volumeToMount.GenerateErrorDetailed("VolumeFSResize.FindAttachablePluginBySpec failed", err)
-	}
-
-	volumeAttacher, err := attachableVolumePlugin.NewAttacher()
-	if err != nil || volumeAttacher == nil {
-		if volumeAttacher == nil {
-			err = fmt.Errorf("VolumeAttacher is nil")
+	fsResizeFunc := func() (error, error) {
+		var resizeDone bool
+		resizeOptions := volume.NodeResizeOptions{
+			VolumeSpec: volumeToMount.VolumeSpec,
 		}
-		return volumetypes.GeneratedOperations{}, volumeToMount.GenerateErrorDetailed("VolumeFSResize.NewAttacher failed", err)
-	}
 
-	deviceMountPath, err := volumeAttacher.GetDeviceMountPath(volumeToMount.VolumeSpec)
-	if err != nil {
-		return volumetypes.GeneratedOperations{}, volumeToMount.GenerateErrorDetailed("VolumeFSResize.GetDeviceMountPath failed", err)
-	}
+		attachableVolumePlugin, _ :=
+			og.volumePluginMgr.FindAttachablePluginBySpec(volumeToMount.VolumeSpec)
 
-	fsResizeFunc := func() (error, error) {
-		resizeSimpleError, resizeDetailedError := og.resizeFileSystem(volumeToMount, volumeToMount.DevicePath, deviceMountPath, volumePlugin.GetPluginName())
+		if attachableVolumePlugin != nil {
+			volumeAttacher, _ := attachableVolumePlugin.NewAttacher()
+			if volumeAttacher != nil {
+				resizeOptions.CSIVolumePhase = volume.CSIVolumeStaged
+				resizeOptions.DevicePath = volumeToMount.DevicePath
+				dmp, err := volumeAttacher.GetDeviceMountPath(volumeToMount.VolumeSpec)
+				if err != nil {
+					return volumeToMount.GenerateError("VolumeFSResize.GetDeviceMountPath failed", err)
+				}
+				resizeOptions.DeviceMountPath = dmp
+				resizeDone, err = og.resizeFileSystem(volumeToMount, resizeOptions, volumePlugin.GetPluginName())
```

Comment:
Can the call and output processing be put into a common function? I see we're doing the same thing for after node publish

Reply:
you mean something like:

```
resizeDone, simpleError, detailedError := og.nodeExpandVolume(volumeToMount,xxxxx)
if simpleError ! = nil {
  return simpleError, detailedError
}
```

doesn't seem to reduce number of lines much.

Reply:
Also L1494 to 1499

Reply:
I would still argue that this will not result in number of lines being reduced by bunch.  Lets say, `resizeFileSystem` itself returned detailed and simple error and resizing is done or not condition:

so
```go
resizeDone, resizeError = og.resizeFileSystem(volumeToMount, resizeOptions, volumePluginName)
			if resizeError != nil {
				klog.Errorf("MountVolume.resizeFileSystem failed with %v", resizeError)
				return volumeToMount.GenerateError("MountVolume.Setup failed while expanding volume", resizeError)
			}
```

Will become:

```go
resizeDone, simpleError, detailedError := og.resizeFilesystemvolumeToMount,xxxxx)
if simpleError ! = nil || detailedError != nil {
  return simpleError, detailedError
}
```

It is literally just one line of saving per-call and loss of context where resizeFilesystem was called from. I am not sure if it is worth it.

Reply:
Sorry I don't mean that, I meant the handling of marking volume as resized.  This logic is duplicated in both places:
```
				if resizeDone {
					markFSResizedErr := actualStateOfWorld.MarkVolumeAsResized(volumeToMount.PodName, volumeToMount.VolumeName)
					if markFSResizedErr != nil {
						// On failure, return error. Caller will log and retry.
						return volumeToMount.GenerateError("VolumeFSResize.MarkVolumeAsResized failed", markFSResizedErr)
					}
					return nil, nil
				}
			}
		}
```

We'll go from 20 lines to 10

Reply:
fixed